### PR TITLE
ODC Layer Implemented

### PR DESCRIPTION
**Pull request by prosrikant**
The New ODC layer is implemented. Mostly the changes are in Requester(New task) and Contributor (search task) screens along with mail changes.

### CODE REVIEW

File: DCode.Common/EmailHelper.cs [hunk 1]
```diff
@@ -53,6 +53,7 @@ public static void SendEmail(string toMailAddress, string ccMailAddress, MailMes
 
                     SmtpServer.Port = 25;
                     SmtpServer.Credentials = new System.Net.NetworkCredential(ConfigurationManager.AppSettings[Constants.DcodeEmailId], ConfigurationManager.AppSettings[Constants.DcodeEmailPwd]);
+                    SmtpServer.UseDefaultCredentials = true;
```
Comment (ritwiksen) [inline DCode.Common/EmailHelper.cs]: @prosrikant :Let's have this removed and leave the email sending mechanism as is.

File: DCode.Web/Scripts/App_JS/Controllers/Contributor.js [hunk 2]
```diff
@@ -105,38 +84,38 @@
...
         };
 
         $scope.filterTasks = function () {
+            //var searchOptions1 = { ProjectName: "" };
             var searchOptions2 = { Task: { ProjectName: null } };
             if ($scope.searchBox != null) {
+                //searchOptions1.ProjectName = $scope.searchBox.text;
                 searchOptions2.Task.ProjectName = $scope.searchBox.text;
             }
+            //$scope.tasks = $filter('filter')(angular.copy($scope.tasksGlobal), searchOptions1);
```
Comment (ritwiksen) [inline DCode.Web/Scripts/App_JS/Controllers/Contributor.js]: Please remove all commented code.

File: DCode.Common/EmailHelper.cs (hunk 1 above)
Comment (prosrikant) [reply DCode.Common/EmailHelper.cs]: Sure. I had put that for testing purpose.

File: DCode.Common/EmailHelper.cs [hunk 3]
```diff
@@ -155,16 +155,24 @@ public static MailMessage ApplyFINotification(string requestorName, string contr
...
             inlineDeloitteLogo.ContentId = Guid.NewGuid().ToString();
             using (var mailMessage = new MailMessage())
             {
                 mailMessage.Subject = string.Format(Constants.DCodeNewFINotification, taskName);
                 mailMessage.IsBodyHtml = true;
-                var textBody = string.Format(Constants.PostNewFIBody, taskName, hours, startDateTime, description, offering);
+                var textBody = string.Empty;
+                if (!string.IsNullOrEmpty(odcMailFooterText) && !string.IsNullOrEmpty(odcMailHeaderImage))
+                {
+                    textBody = string.Format(Constants.PostNewFIBody, taskName, hours, startDateTime, description, offering, odcMailFooterText);
```
Comment (ritwiksen) [inline DCode.Common/EmailHelper.cs]: Please use a ternary operator to evaluate the footer text.

File: DCode.Services/Common/CommonService.cs [hunk 4]
```diff
@@ -934,5 +946,55 @@ private string GetLocationForUser(string userName)
             }
             return null;
         }
+
+        /// <summary>
+        /// This method is responsible to indentify if the user is part of any specific ODC's distribution list.
+        /// If yes he will have access to additional feature specific to that ODC
+        /// </summary>
+        /// <returns></returns>
+        private void ValidateAndSetODCAccess(UserContext _userContext)
```
Comment (ritwiksen) [inline DCode.Services/Common/CommonService.cs]: Let's not have the Common Service working on functionality that is specific to the `ExchangeUser`. Can we move all `ExchangeUser` functionality to an independent service altogether. Also, let's return a boolean out of this method as we're trying to validate some logic in place of returning a void.

File: DCode.Services/Common/ODCReferenceService.cs [hunk 5]
```diff
@@ -0,0 +1,79 @@
...
+            var odcList = GetExistingODCList(xmlFilePath);
+            return odcList.ODCList.Where(off => off.OfferingId == offeringId).FirstOrDefault();
+        }
+    }
+
+    /// <summary>
+    /// Individual ODC node containing the details of an ODC
+    /// </summary>
+    [Serializable]
+    public class ODC
```
Comment (ritwiksen) [inline DCode.Services/Common/ODCReferenceService.cs]: Please have all POCO's as independent files.

File: DCode.Services/Reporting/ReportingService.cs [hunk 6]
```diff
@@ -95,7 +95,7 @@ public IEnumerable<string> GetSkillsForNewTasksAddeddYesterday()
 
         public void UpdateDailySiteVisitCount()
         {
-            _dailyUsageStatisticsRepository.UpsertDailyStatistics();
+            //_dailyUsageStatisticsRepository.UpsertDailyStatistics();
```
Comment (ritwiksen) [inline DCode.Services/Reporting/ReportingService.cs]: Please have this code uncommented as we're tracking user hits count.

File: DCode.Web/Content/all.css [hunk 7]
```diff
@@ -1544,16 +1560,16 @@ input.edit-task-btn.width110 {
     }
 
 .new-style-DTM-create-btn-side {
-    width: 33%;
+    width: 20%;
```
Comment (ritwiksen) [inline DCode.Web/Content/all.css:1563]: Will this impact any component in the existing site? The same queries for the other style updates as well.

File: DCode.Web/Scripts/App_JS/Controllers/Contributor.js [hunk 8]
```diff
@@ -105,38 +84,38 @@
                 $scope.divVisibiltyModel.showSummary = true;
                 $scope.divVisibiltyModel.showSuccess = false;
                 $scope.reviewIndex = index;
+                //document.getElementById('divManagerEmailId' + index).get(0).focus();
                 setTimeout(function () { $('#txtManagerEmailId' + index).focus() }, 1);
-
+                //$location.hash('div' + index);
```
Comment (ritwiksen) [inline DCode.Web/Scripts/App_JS/Controllers/Contributor.js]: Please remove all commented lines of code.

File: DCode.Web/Scripts/App_JS/Controllers/Contributor.js [hunk 9]
```diff
@@ -225,11 +203,31 @@
...
+                                value.isODCTask = false;
+
+                                if (value.SubOfferingId != null && value.SubOfferingId != "" && value.SubOfferingId != 0) { //i.e. -> The task belongs to an ODC
+                                    if (value.OfferingId == $scope.accessibleODCId) { //The Person has access to the ODC
+                                        value.isODCTask = true;
+                                        tempArray.push(value);
+                                    }
+                                } else if (!$scope.showOnlyODCTask && $scope.showOnlyODCTask != undefined) //--chararora@deloitte.com, to check the value of the checkbox
+                                {
+
```
Comment (ritwiksen) [inline DCode.Web/Scripts/App_JS/Controllers/Contributor.js]: Please remove empty lines.

File: DCode.Web/Scripts/App_JS/Controllers/Contributor.js [hunk 10]
```diff
@@ -328,6 +330,41 @@
...
+                method: "GET"
+            }).success(function (data, status, config) {
+                if (data != null) {
+                    $scope.ODCList = data.ODCList;
+                    for (var i = 0; i < data.ODCList.length; i++) {
+                        if (data.ODCList[i].OfferingId == $scope.accessibleODCId) {
+                            $scope.accessibleODCName = data.ODCList[i].Name;
+                        }
+                    }
+                    console.log(data);
```
Comment (ritwiksen) [inline DCode.Web/Scripts/App_JS/Controllers/Contributor.js]: Please remove all `console.logs`.

File: DCode.Web/Views/Shared/_Layout.cshtml [hunk 11]
```diff
@@ -230,7 +230,7 @@
                 <div class="new-style-DTM-layout-logo col-sm-10">
                     @*<div style="height:200px; width:200px">*@
                     <div @*style="float:left;width:80%;"*@>
-                        <img src="~/Content/Images/tech-x-logo.png" class="new-style-DTM-layout-logo-image pull-left logo">
+                        <img ng-src="~/Content/Images/{{logoImageName}}" class="new-style-DTM-layout-logo-image pull-left logo">
```
Comment (ritwiksen) [inline DCode.Web/Views/Shared/_Layout.cshtml]: Could we have an alt text as the image would be lazy loaded till angular loads the value of the image path.

**Review by ritwiksen — CHANGES_REQUESTED** [review]

File: DCode.Common/EmailHelper.cs (hunk 3 above)
Comment (prosrikant) [reply DCode.Common/EmailHelper.cs]: done

File: DCode.Services/Common/CommonService.cs (hunk 4 above)
Comment (prosrikant) [reply DCode.Services/Common/CommonService.cs]: made the method to SetODCAccess and moved it to ODCService

File: DCode.Services/Reporting/ReportingService.cs (hunk 6 above)
Comment (prosrikant) [reply DCode.Services/Reporting/ReportingService.cs]: done

File: DCode.Web/Scripts/App_JS/Controllers/Contributor.js (hunk 8 above)
Comment (prosrikant) [reply DCode.Web/Scripts/App_JS/Controllers/Contributor.js]: done.. I guess, this was commented before.

File: DCode.Web/Scripts/App_JS/Controllers/Contributor.js (hunk 9 above)
Comment (prosrikant) [reply DCode.Web/Scripts/App_JS/Controllers/Contributor.js]: done

File: DCode.Web/Scripts/App_JS/Controllers/Contributor.js (hunk 10 above)
Comment (prosrikant) [reply DCode.Web/Scripts/App_JS/Controllers/Contributor.js]: done

File: DCode.Services/Common/ODCReferenceService.cs (hunk 5 above)
Comment (prosrikant) [reply DCode.Services/Common/ODCReferenceService.cs]: implemented separate classes

File: DCode.Web/Content/all.css (hunk 7 above)
Comment (prosrikant) [reply DCode.Web/Content/all.css:1563]: No this does not impact

File: DCode.Web/Views/Shared/_Layout.cshtml (hunk 11 above)
Comment (prosrikant) [reply DCode.Web/Views/Shared/_Layout.cshtml]: done

File: DCode.Web/Scripts/App_JS/Controllers/Contributor.js (hunk 2 above)
Comment (prosrikant) [reply DCode.Web/Scripts/App_JS/Controllers/Contributor.js]: done

**Review by ritwiksen — APPROVED** [review]

**Pull request merged**